vowpalwabbit/csv_parser/include/vw/csv_parser/parse_example_csv.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#pragma once
+
+#include "vw/common/text_utils.h"
+#include "vw/core/global_data.h"
+#include "vw/core/v_array.h"
+
+#include <map>
+#include <vector>
+
+namespace VW
+{
+namespace parsers
+{
+namespace csv
+{
+int csv_to_examples(VW::workspace* all, io_buf& buf, VW::v_array<example*>& examples);
+
+class parser
+{
+public:
+  parser() = default;
+
+  int parse_csv(VW::workspace* all, io_buf& buf, VW::example* ae);
+
+private:
+  std::vector<std::string> _header;
+  uint64_t _channel_hash;
```

Comment:
Yeap, I think a special header in the CSV file which allows you to specify the ns would be great. Options to specify which columns belong to which namespaces would be overly complexing for users.

Take the Iris2.csv as an example, we may use the first `.` as a seperator.
`sepal.length,sepal.width,petal.length,petal.width`

the string before the first `.` will be the namespace, after that will be the feature name. If no `.` is present, the ns will be empty. 

Users may also have the option to specify the which separator to use to avoid the collision with the csv separator.

Reply:
Will support the namespace feature later. Currently use empty namespace as the replacement.

Reply:
Now this gets supported in my last commit.

> Yeap, I think a special header in the CSV file which allows you to specify the ns would be great. Options to specify which columns belong to which namespaces would be overly complexing for users.
> 
> Take the Iris2.csv as an example, we may use the first `.` as a seperator. `sepal.length,sepal.width,petal.length,petal.width`
> 
> the string before the first `.` will be the namespace, after that will be the feature name. If no `.` is present, the ns will be empty.
> 
> Users may also have the option to specify the which separator to use to avoid the collision with the csv separator.

Reply:
In VW `sepal.length` is a valid feature name and valid namespace. The only character we can use here is a `|` (this is due to the very flexible text format). This was another reason why I was sorted of pushing for not allowing arbitrary separators. I guess we could support escaping here, but thats another complexity.

Question do you intend for this information to be a part of the normal csv header or be an additional header.
Would it be:
```csv
sepal.length,sepal.width,petal.length,petal.width
length, width, length, width
1,2,3,4
```

Or just:
```csv
sepal.length,sepal.width,petal.length,petal.width
1,2,3,4
```

Reply:
> The only character we can use here is a `|` (this is due to the very flexible text format)

OK, I will change the default namespace separator to `|` to make it compatible with VW format. (I also make an option available here to specify the arbitrary separators for namespaces, as `.` can really be a valid namespace separator for `sepal.length` )

> This was another reason why I was sorted of pushing for not allowing arbitrary separators. I guess we could support escaping here, but thats another complexity.

Here's how I'm handling this. the CSV separator has a higher priority than the namespace separator, e.g. when they are both specified as `|`, there will be no namespace to separate. As users has plenty of valid characters to choose as the separator, there's no need to escape. 

I'm intended for this information to be a part of the normal csv header, as we don't have to break the CSV format for this feature, it's just:
```csv
sepal.length,sepal.width,petal.length,petal.width
1,2,3,4
```

Reply:
What is a valid scenario where namespace separator and csv separator are the same? I think it should be an error if namespace separator and csv separator are the same.

Reply:
> What is a valid scenario where namespace separator and csv separator are the same? I think it should be an error if namespace separator and csv separator are the same.

Since the CSV separator has a higher priority than the namespace separator, if they are the same, we would assume that the header doesn't include the namespace information. An error will be thrown if the separated header parts number is greater than the values parts', which always occurs if the header includes the namespace information and when the namespace separator and csv separator are the same.

Reply:
To me that is confusing though. It doesn't seem like a valid scenario, but instead something that just happens to work.

Put another way, if we don't allow namespace separator and csv separator to be the same what situation would be break?

Reply:
> Put another way, if we don't allow namespace separator and csv separator to be the same what situation would be break?

Since now we make `|` the default for namespace separator, if user specify the CSV separator to be `|` while not explicitly specify anything for namespace separator, which users may mean that the header doesn't include the namespace information, and the namespace separator is not needed at all, thrown an error would be a trouble to them, as that doesn't seems like what an intelligent parser would do.

To avoid confusion, maybe we can add a warning to warn such situation and notify users what we'll going do next when namespace separator and csv separator are the same. Throwing an error is not needed as we can handle such a situation.

Reply:
Saying that the csv separator has higher priority has the same meaning as saying when csv and namespace separators are the same value then ALL namespace separators are ignored.

If there is no valid reason to do this, then why let it be invalid input? The notion of priority adds cognitive load to users specifying options. They need to keep in their mind which options override which other options to understand what will end up happening. By disallowing the same value being used we are removing the concept of priority which I believe is a usability improvement.

Reply:
Okay. so will change the warning into error.